### ISSUE TITLE
Implement optional joystick lock for raw data tooltip and disable secondary override on movement for other tooltips

### ISSUE THREAD

**Issue by Crusher50** [Yellow-Dog-Man/Resonite-Issues#125]
### Is your feature request related to a problem? Please describe.

When a tooltip is equipped on a controller that does not have an extra touch pad for the secondary action, so the joystick click is used instead (i.e. every controller that isn't Valve Index controllers) the locomotion is locked to one controller and becomes tank controls. This can be very disorientating and cumbersome when developing with tool tips. 

Another issue this causes is when two tooltips are equipped in both hands the user becomes stuck, as both joysticks movements are locked. This can, again, be annoying when developing.

I also understand that grabbing an object would need to lock locomotion on non-index controllers. I have only made this issue in regards to secondary actions with a tooltip equipped.

### Describe the solution you'd like

There are many possible solutions to this issue, here are just a few:

The first is a suggestion provided by @shadowpanther from a comment on the previous platform.

"Personally, I'm in favour of separating "use secondary" and "use joystick axes" into two separate options. That might require to update some older user-made tools, but also that new property might be made to copy the old behaviour on older schema load, and the default tools updated to not require axes specifically."

I too believe this would be the best solution but if this doesn't meet your requirements, I have some more solutions:

### Swap the secondary and dash buttons around on non-index controllers by default:
- This is the simplest solution I can think of in this moment in time. If this were to be implemented then tooltips that use secondary would no longer require the joystick to be pressed down, while still retaining full locomotion as there is no longer any reliance on the joystick. 
- This *in theory* should also reduce the amount of wear on joysticks as the joystick would not be pressed down as often (At least for people who regularly use tools). 
- **Possible downsides**:
- The only possible downsides to this solution I can think of is accidental miss clicks for some users that apply pressure to joysticks while locomoting. However, I believe that accidental secondary presses could possibly be more harmful to others (stray gizmos, custom tool functionality, etc) in a given session then accidently opening the dash. This may not be an issue for people who do not build often in Resonite but until controller input settings (issue number TBD) is implemented this would be the better alternative for the vast majority of Resonite users. 
- Some tools may require the joystick axes to function. However, for the default tool set that are not custom (dev-tip, flux-tip, etc) the functionally could be altered to not override locomotion, as in testing they do not require the joystick axes to function. 

### Give people the option to disable locomotion override in the settings menu, even if it may hinder some tools functionality:
- This would require the settings UI rework, but would give users the most freedom in choice for the user's desired preference no matter the skill level. This can be disabled by default, with a warning of possible hindered functionality, but at the very least giving people the option to change this functionally at all is still better then the current situation.
- **Possible downsides**:
- If people do disable the locomotion override some tools that require the joystick axis for functionality may be affected. 

### Add an option to override locomotion in the raw data tool tip component (and other tool tips that are applicable) and disable it by default on all current tools that do not require this functionality:
- This would allow creators to decide whether their tools requires the joystick to work in the absence of a touch pad that is used on index controllers. This can, kind of, already be done by disabling secondary, but this then disables the secondary functionality all together which isn't a viable solution in some instances.
- **Possible downsides**:
- I honestly don't see any major downsides to this solution then some tools not functioning the same way as users may typically expect. (although this already happens with users expecting the same locomotion they are used to on index when they swap to any other headset that is not compatible with index controllers).
- There may also be a lack of documentation for users that are unaware of this option and thus may get stumped when trying to find a solution. ( Although this already happens when users try to find a solution to disable tank controls in the first place.)

### Describe alternatives you've considered

Currently you can disable secondary usage on a raw data tooltip, but this is only for raw data tooltips, and disables the ability to use secondary.

Otherwise, there is a mod made by @furrz that can be found here: https://github.com/furrz/NoTankControls that fixes this issue.

### Additional Context

- Previously this feature request / issue was taken as by design and was not applicable to a fix, and that there were no viable alternative (except for a compromise, but I feel the solutions above are a far better alternative). 

- Hopefully this is enough to show that there are other viable alternatives (and many more that I can list but this issue is long enough as is). 

- Until then I, and many others, will most likely be using community made solutions, like the mod listed above.

**Comment by Stellanora64** [comment]
I'm curious as to how the mod even functions then, as by this description it seems that joystick click and direction are considered one in the same by resonite.

**Comment by Stellanora64** [comment]
@shiftyscales Understandable, however I still don't see a reason as to separate the secondary action and secondary axis out into two separate toggles in Rawdata tooltip. As in this case custom tools already have no consistency in joystick lock due to disabling secondary, but will allow creators to give the user the ability to use their joystick for movement (for exampe for a weapon that needs reload would require secondary, but also requires better movement for a shooter game). 

This could also be applied to the normal tools, and then only be enabled when it is required, but as your stance seems very apposed to this they could be left out. 

But if there is another reason for this not being possible,  I hope that at least some other solution could be investigated in the future as the quest controller layout becomes more standard.

**Comment by shadowpanther** [comment]
There still needs to be a way, I think, to decouple stick axes from the secondary so that the tools only use axes if they need axes, and only use the stick click if they only need the secondary as a button.

Default tools _right now_ do not require analog axes on the secondary, as far as I know. Therefore, these tools do not _need_ to block the movement controls.

Otherwise this heavily inconveniences Oculus/Meta users's use of tools. Not everyone has enough money and play space for a Vive/Index setup.

**Comment by shadowpanther** [comment]
> #23 could be a possible resolution to the issue- but it would be on the end-user to find something else they can bind their secondary axis to for tools that require it.

Rebinding would not allow to fix this issue as there are too few buttons on a Quest-style controller. Everything is already used.

For example, say I want to transfer even one secondary input to another button so that the stick is always my movement/jump. What other function should I sacrifice?
* Context menu? Tools often use it, so I need it on both hands.
* Dash? Both Dash buttons are used to enable the Dash edit mode.
* Grip of Trigger? Both are essential interaction inputs.
* App menu? Is used by SteamVR.
* Meta menu? Is used by Quest itself.

See, I'm out of rebinding options already. And I didn't even try to use two tools at the same time, which would require two secondary buttons. So, how does the rebinding help in my case?

On the other hand, allowing for the _tool_ to release the axes while keeping the secondary _button_ would help a lot until the tool would _actually_ require axes for its functionality in its next update.

**Comment by shadowpanther** [comment]
Compromise - what if there would be an _option_ to decouple the axes from the secondary? I would willingly sacrifice the functionality of the few tools that currently use axes and keep my locomotion. When the default tools start using axes, I might turn this option off, but that might not happen for quite some time.

This would not require to change how the tools work, only to optionally separate the stick axes into a separate input that is not blocked by the tools using a secondary.

**Comment by Frooxius** [comment]
> Compromise - what if there would be an option to decouple the axes from the secondary?

This would make things even way more complicated on our end. Now when designing things, we have to worry about usability of the tool with that setting on and with that setting off, rather than just dealing with a single consistent system.

I think the main thing here would be going over some of the basic principles of the tools and removing need to use the Secondary on a lot of them.

The biggest problem in decoupling them is that it won't work for all controllers - e.g. Vive wands, you cannot reliably use Secondary without clicking it.

**Comment by shadowpanther** [comment]
> I think the main thing here would be going over some of the basic principles of the tools and removing need to use the Secondary on a lot of them.

While this sounds good, consider two main developer tools - Dev Tool and Flux Tool. I don't see any way the need for the secondary could be removed from them without complicating their use. Like, move all interactions into laser+context territory? Use primary instead of the secondary?

And with Shifty speaking of _adding_ the need for axes in the future instead of _removing_ the need for the secondary on the tools, two stories don't quite fit together.

> The biggest problem in decoupling them is that it won't work for all controllers - e.g. Vive wands, you cannot reliably use Secondary without clicking it.

I guess that option would not be popular with Vive wands users, then. But an option is still better than no option.

**Comment by Stellanora64** [comment]
> When default tools do get updated/include the use of the secondary axis as part of their function- that just brings us back to square one, @shadowpanther @Stellanora64.
> 
> #23 could be a possible resolution to the issue- but it would be on the end-user to find something else they can bind their secondary axis to for tools that require it.

For the default tools that makes sense then. However, this doesn't answer my question for not adding an option to the RawDataTooltip component?

I could see #23 being a good solution. 


> > Compromise - what if there would be an option to decouple the axes from the secondary?
> 
> This would make things even way more complicated on our end. Now when designing things, we have to worry about usability of the tool with that setting on and with that setting off, rather than just dealing with a single consistent system.

I assume this is possibly the reason not adding the option for RawDataTooltip? That being the RawDataTooltip, and the default tool system are a shared system (correct me if I'm wrong.)

**Comment by Frooxius** [comment]
> For the default tools that makes sense then. However, this doesn't answer my question for not adding an option to the RawDataTooltip component?

It already has that option. Uncheck the `UseSecondary` toggle.

**Comment by shadowpanther** [comment]
> > For the default tools that makes sense then. However, this doesn't answer my question for not adding an option to the RawDataTooltip component?
> 
> It already has that option. Uncheck the `UseSecondary` toggle.

They talk about the other option, see the quote of myself from the old days:

> "Personally, I'm in favour of separating "use secondary" and "use joystick axes" into two separate options. That might require to update some older user-made tools, but also that new property might be made to copy the old behaviour on older schema load, and the default tools updated to not require axes specifically."

**Comment by Frooxius** [comment]
In that case that would been to be a global change. RawDataTooltip just follows the principles of the general tooltip system, which doesn't have this feature.

**Comment by shadowpanther** [comment]
Could this feature be implemented and then the default tools updated to not require the axes (provided, they do not require the axes in the first place)?

**Comment by Stellanora64** [comment]
> > For the default tools that makes sense then. However, this doesn't answer my question for not adding an option to the RawDataTooltip component?
> 
> It already has that option. Uncheck the `UseSecondary` toggle.

Apologies for the confusion, mean to have secondary axes and secondary action be separated on rawdatatooltip. shadowpanther is right for what I mean. 

Given your response though, it seems that my suspension was correct in that both the default tools and Rawdata tooltip share the same system, and to update one would mean updating the other.
